src/io/cmap.rs

```diff
@@ -0,0 +1,1038 @@
+            }
+            if buffer.ends_with("\t0")
+                || buffer.ends_with("\t0\n")
+                || buffer.ends_with("\t0\r\n")
+                || buffer.is_empty()
+            {
+                break;
+            }
+        }
+        if num_channels == 0 {
+            bail!(Error::InvalidLabelChannel("1"))
+        } else {
+            Ok(Reader {
+                stream,
+                header,
+                buffer,
+            })
+        }
+    }
+}
+
+impl<R: BufRead> CmapRead for Reader<R> {
+    fn read_into(&mut self, record: &mut Record) -> Result<bool> {
+        if self.buffer.is_empty() {
+            return Ok(false);
+        }
+        let buf_lines: Vec<&str> = self.buffer.trim().lines().collect();
+        let mut res = match record.fill(buf_lines) {
+            Ok(()) => Ok(true),
+            Err(e) => Err(e),
+        };
+        self.buffer.clear();
+        loop {
+            res = match self.stream.read_line(&mut self.buffer) {
+                Ok(0) => {
+                    if self.buffer.is_empty() {
+                        res
+                    } else {
+                        bail!(Error::IncompleteCmapRecord)
+                    }
+                }
+                Ok(_) => res,
+                Err(e) => Err(e.into()),
+            };
+            if self.buffer.ends_with("\t0")
+                || self.buffer.ends_with("\t0\n")
+                || self.buffer.ends_with("\t0\r\n")
+                || self.buffer.is_empty()
+            {
+                break;
+            }
+        }
+        res
+    }
+}
+
+/// Iterator over records.
+impl<R: BufRead> Iterator for Reader<R> {
+    type Item = Result<Record>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut rec = Record::new();
+        match self.read_into(&mut rec) {
+            Ok(true) => Some(Ok(rec)),
+            Ok(false) => None,
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+/// A container with CMAP header and records.
+#[derive(Debug, new, PartialEq)]
+pub struct Container {
+    header: Vec<String>,
+    contigs: HashMap<u32, Record>,
+}
+
+impl Container {
+    /// Constructs a new `Container` from path.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let reader = Reader::from_path(path)?;
+        reader.into_container()
+    }
+
+    /// Returns header.
+    pub fn header(&self) -> &Vec<String> {
+        &self.header
+    }
+
+    /// Returns label positions (for each contig respectively).
+    pub fn labels(&mut self) -> HashMap<u32, Vec<f64>> {
+        self.contigs
+            .clone()
+            .iter_mut()
+            .map(|(id, rec)| {
+                (
+                    *id,
+                    rec.labels()
+                        .iter()
+                        .map(|label| label.pos)
+                        .collect::<Vec<_>>(),
+                )
+            })
+            .collect::<HashMap<_, _>>()
+    }
```

Comment:
Instead of creating a new hashmap on the fly, I would rather suggest a method

```rust
pub fn record(&self, contig: &[u8]) -> &Record
```

Reply:
I don't understand really what you mean by this. Why do we return a `Record` here?

Reply:
I added the new function but kept `labels()` for now.